addons/account/wizard/account_merge_wizard.py

```diff
@@ -0,0 +1,319 @@
+            raise UserError(_("This can only be used on accounts."))
+
+        res['account_ids'] = [Command.set(self.env.context.get('active_ids'))]
+        return res
+
+    @api.model
+    def _get_grouping_key(self, account, group_by_name=False):
+        """ Return a grouping key for the given account.
+        If you modify this method, make sure you modify `_get_group_display_name` correspondingly.
+        """
+        self.ensure_one()
+        grouping_fields = ['account_type', 'non_trade', 'currency_id', 'reconcile', 'deprecated']
+        if group_by_name:
+            grouping_fields.append('name')
+        return tuple(account[field] for field in grouping_fields)
+
+    @api.model
+    def _get_group_display_name(self, grouping_key, group_by_name=False):
+        """ Return a human-readable name for the group, in the format:
+        '{Trade/Non-trade} Receivable {USD} {Reconcilable} {Deprecated}'
+        """
+        self.ensure_one()
+
+        account_type = grouping_key[0]
+        account_type_label = dict(self.env['account.account']._fields['account_type']._description_selection(self.env))[account_type]
+        if account_type in ['asset_receivable', 'liability_payable']:
+            account_type_label = _("Non-trade %s", account_type_label) if grouping_key[1] else _("Trade %s", account_type_label)
+
+        other_key_elements = []
+        if grouping_key[2]:
+            other_key_elements.append(self.env['res.currency'].browse(grouping_key[2]).name)
+
+        if grouping_key[3]:
+            other_key_elements.append(_("Reconcilable"))
+
+        if grouping_key[4]:
+            other_key_elements.append(_("Deprecated"))
+
+        if not group_by_name:
+            grouping_key_name = account_type_label
+            if other_key_elements:
+                grouping_key_name = f'{grouping_key_name} ({", ".join(other_key_elements)})'
+
+        else:
+            grouping_key_name = f'{grouping_key[5]} ({", ".join([account_type_label] + other_key_elements)})'
+
+        return grouping_key_name
+
+    @api.depends(
+        'is_group_by_name',
+        'account_ids.name',
+        'account_ids.account_type',
+        'account_ids.non_trade',
+        'account_ids.currency_id',
+        'account_ids.reconcile',
+        'account_ids.deprecated',
+    )
+    def _compute_wizard_line_ids(self):
+        """ Determine which accounts to merge together. """
+        for wizard in self:
+            # Filter out Bank / Cash accounts
+            accounts = wizard.account_ids._origin.filtered(lambda a: a.account_type not in ('asset_bank', 'asset_cash'))
+            accounts_grouped = accounts.grouped(lambda a: wizard._get_grouping_key(a, group_by_name=wizard.is_group_by_name))
+
+            wizard_lines_vals_list = []
+            sequence = 0
+            for grouping_key, group_accounts in accounts_grouped.items():
+                grouping_key_str = str(grouping_key)
```

Comment:
I agree, ideally, I would want to have something like a grouped list view in the one2many widget. Unfortunately that's not available, and I don't know how complicated it would be to implement it. I agree though that doing the grouping server-side feels kind of clunky.

In any case, at the moment, the values in the `grouping_key` field are only necessary only in order to recompute the `info` group-by-group, so arguably, we don't even need a field for it.

Reply:
Actually, having a grouping_key field simplifies the computes a lot, so I'll just keep it for now - but do let me know if you find something that works better!